feat: Layout conversion work flow

rahulramesha opened the pull request:
# Description

This PR contains changes to provide the user with a way to convert between auto layout and fixed layout while also creating a snapshot of the app to provide a safety net. The changes include,
-  Auto to fixed Conversion Algorithm (Additional description below)
- Fixed to Auto Conversion Algorithm (Is pretty much straight forward and uses existing calculations)
- API and sagas to create, fetch, delete and Restore snapshot
- API and saga to save all the pages of the application
- Dialog with Conversion Form which takes in flow to convert layouts/ restore snapshot
- Hooks that define each flow state for conversion from Auto to fixed, fixed to Auto and restoring snapshot
- Banner Message on top of canvas to indicate the status of snapshot

The Fixed to Auto conversion algorithm works as below,
- If the Widget is a CANVAS_WIDGET, it is forwarded to process the positions of children and fit them into Flex Layers
- If the widget is not a CANVAS_WIDGET only it's property and positions gets updated
- For CANVAS_WIDGET, it's children are first sorted by topRow
- Of the Children, the top most widget is taken and it is used to find the left most widget along its rows
- With the left Top most widget, It is used to find out the next widget in the layer by checking if any widgets are along it's rows to the right of the widget
- By finding next widget in layer, that particular widget is filtered out of the children from next iteration
- While fitting the widgets into layers, the widgets are grouped if the space between widgets is more than 10% of the total width of canvas
- If number of groups are more than 3, it is condensed into 3 distinct groups
- Based on these groups, the group's Alignment score is calculated and the widgets inside the group are aligned accordingly
- A similar scoring method is used to determine the vertical Alignment of the Widgets within the layer.
- Algorithm ends when there are no more children left to process.


Fixes #19724
Fixes #19726
Fixes #19826
Fixes #20264 
Fixes #19590
Fixes #20990 
Fixes #20988 
Fixes #20989
Fixes #20987 
Fixes #20595
Fixes #20986 
Fixes #19512
Fixes #19693

Media
> A video or a GIF is preferred. when using Loom, don’t embed because it looks like it’s a GIF. instead, just link to the video


## Type of change
- New feature (non-breaking change which adds functionality)


## How Has This Been Tested?
- Manual
- Jest
- Cypress

### Test Plan
> Add Testsmith test cases links that relate to this PR

### Issues raised during DP testing
> Link issues raised during DP testing for better visiblity and tracking (copy link from comments dropped on this PR)


## Checklist:
### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

rahulramesha commented:
/build-deploy-preview

rahulramesha commented:
/build-deploy-preview

riodeuno commented on app/client/src/actions/autoLayoutActions.ts:
```diff
@@ -18,6 +23,34 @@ export const updateLayoutForMobileBreakpointAction = (
+  };
+};
+
+export const updateSnapshotDetails = (
```
In which scenario would the snapShotDetails be undefined? Is it necessary to update the snapShotDetails if it is undefined?

riodeuno commented on app/client/src/api/ApplicationApi.tsx:
```diff
@@ -209,6 +209,11 @@ export interface PageDefaultMeta {
 
+export interface snapShotApplicationRequest {
+  applicationId: string;
+  branchName?: string;
```
We are never using the branchName in the API calls. Is it going to be implemented in the future?

rahulramesha replied:
missed it, removed.

riodeuno commented on app/client/src/api/ApplicationApi.tsx:
```diff
@@ -338,6 +343,33 @@ class ApplicationApi extends Api {
   }
+
+  static createSnapShotOfApplication(request: snapShotApplicationRequest) {
+    return Api.post(
```
I was wondering why this is a POST request? Are we planning to send any data from the client to help with the snapshots? Is this going to be the branchName we've configured in the type above?

riodeuno replied:
Also, if we map this to the standard REST nomenclature, shouldn't this be "PUT" instead?

riodeuno commented on app/client/src/api/ApplicationApi.tsx:
```diff
@@ -338,6 +343,33 @@ class ApplicationApi extends Api {
+    );
+  }
+
+  static restoreSnapShotOfApplication(request: snapShotApplicationRequest) {
```
Would a better name here be restoreApplicationFromSnapshot? I'm asking because, this name, tells me more about about what is happening. Please let me know if I am misunderstanding something here.

rahulramesha replied:
changed

riodeuno commented on app/client/src/api/ApplicationApi.tsx:
```diff
@@ -338,6 +343,33 @@ class ApplicationApi extends Api {
+    );
+  }
+
+  static deleteSnapShotOfApplication(request: snapShotApplicationRequest) {
```
We can also get rid of the `ofApplication` suffix in the names, it doesn't provide more context than what we can already gather from the file name. Do we have other snapshots in the ecosystem?

rahulramesha replied:
changed

riodeuno commented on app/client/src/api/PageApi.tsx:
```diff
@@ -213,6 +225,15 @@ class PageApi extends Api {
     );
   }
 
+  static saveAllPages(
```
Thinking out loud here. Would this API call help us with the migrations as well? Today, we migrate one page at a time, we can use this to migrate all pages.

rahulramesha replied:
yeah, it should be possible

riodeuno commented on app/client/src/ce/constants/ApiConstants.tsx:
```diff
@@ -41,3 +41,6 @@ export const SUPER_USER_SUBMIT_PATH = `${SIGNUP_SUBMIT_PATH}/super`;
 export const getExportAppAPIRoute = (applicationId: string) =>
   `/api/v1/applications/export/${applicationId}`;
+
+export const getSnapShotAppAPIRoute = (applicationId: string) =>
```
Can we re-use this in the ApplicationApi file above. We're generating the same value in a different way over there.

riodeuno commented on app/client/src/ce/constants/ReduxActionConstants.tsx:
```diff
@@ -761,6 +761,16 @@ export const ReduxActionTypes = {
   AUTOLAYOUT_ADD_NEW_WIDGETS: "AUTOLAYOUT_ADD_NEW_WIDGETS",
   RECALCULATE_COLUMNS: "RECALCULATE_COLUMNS",
+  UPDATE_LAYOUT_POSITIONING: "UPDATE_LAYOUT_POSITIONING",
+  SET_LAYOUT_CONVERSION_STATE: "SET_LAYOUT_CONVERSION_STATE",
```
What states are there in the layout conversion? Is this the same as conversion init, success and error?

rahulramesha replied:
These are states of the conversion form. So, conversion form is rendered based on the properties returned by the hook based on this state.

riodeuno commented on app/client/src/ce/constants/messages.ts:
```diff
@@ -1460,6 +1461,50 @@ export const SAVE_BUTTON_TEXT = () => "SAVE";
 export const DISCARD_POPUP_DONT_SAVE_BUTTON_TEXT = () => "DON'T SAVE";
 
+//Layout Conversion flow
+export const CONVERT = () => "CONVERT";
```
This is nice! We're using constants for messaging and it will help us in the future.

riodeuno commented on app/client/src/ce/pages/Applications/index.tsx:
```diff
@@ -749,6 +750,9 @@ export function ApplicationsSection(props: any) {
                         INVITE_USERS_MESSAGE,
                         cloudHosting,
                       )}
+                      onOpenOrClose={(isOpen) =>
```
Will it help if we useCallback here?

riodeuno commented on app/client/src/constants/Colors.tsx:
```diff
@@ -164,6 +164,7 @@ export const Colors = {
   GREY_900: "#191919",
 
   PRIMARY_ORANGE: "#F86A2B",
+  PRIMARY_ORANGE_OPAQUE: "#F86A2B1A",
```
Can we confirm if all browsers understand this extended format hex colors?

riodeuno commented on app/client/src/pages/Editor/CanvasLayoutConversion/ConversionCompleteLayout.tsx:
```diff
@@ -0,0 +1,41 @@
+  errorText?: string;
+};
+
+export const ConversionCompleteLayout = (
```
This seems like something which should be coming from the design system instead. Can we confirm if it makes more sense to get this from the design system?

rahulramesha replied:
ConversionCompleteLayout is specific to this, but will create a task to check if this and InfoBlock component can be used in design system instead. It might need some refactor cuz design system doesn't have tailwind css used in the components.

riodeuno replied:
Yes, it may need some refactor. The reason I would like for us to insist on this, is because we will be standardising components across the board, and changes to the Appsmith design system can easily be propagated to this component without having to change anything here.

riodeuno commented on app/client/src/selectors/autoLayoutSelectors.tsx:
```diff
@@ -56,6 +63,62 @@ export const getParentOffsetTop = (widgetId: string) =>
+      ExpirationHours - hoursPassedSince,
+    );
+
+    const readableDate = moment(lastUpdatedDate).format("Do MMMM, YYYY h:mm a");
```
Let's avoid the use of moment and start using dayjs. It has a lot of utilities which can help us avoid all the time manipulations.

riodeuno commented on app/client/src/utils/WidgetFactory.tsx:
```diff
@@ -223,6 +223,15 @@ class WidgetFactory {
     return Array.from(this.widgetMap.keys());
   }
 
+  static getWidgetConfigMap(widgetType: WidgetType): Partial<WidgetProps> {
```
Did we not have this already? 🤔

rahulramesha replied:
My bad, didn't notice the map directly accessible from the WidgetFactory Object. Removed this.

riodeuno commented on app/client/src/utils/autoLayout/AutoLayoutUtils.ts:
```diff
@@ -332,6 +332,7 @@ export function getAlignmentColumnInfo(
       [FlexLayerAlignment.Start]: 0,
       [FlexLayerAlignment.Center]: 0,
       [FlexLayerAlignment.End]: 0,
+      [FlexLayerAlignment.None]: 0,
```
What does None signify?

riodeuno commented on app/client/src/utils/autoLayout/flexWidgetUtils.ts:
```diff
@@ -1,7 +1,7 @@
-  return isMobile && widget.mobileRightColumn !== undefined
-    ? widget.mobileRightColumn
-    : widget.rightColumn;
+  return isMobile && widget?.mobileRightColumn !== undefined
```
So this makes it possible that when isMobile is true, the widget is undefined? Could we elaborate on when this happens?

riodeuno replied:
Also, does typescript not complain, as according to the code, the return could be number | undefined, instead of just number.

riodeuno reviewed (commented):
This is quite complex. I appreciate the fact that we've used a lot of standard practices in the organization of the code. I have a lot of comments which are nitpicks, so please reply in case you think that the comments don't make sense, and things should remain as is.

Also, please add code comments where ever you can, function descriptions are necessary, the rest is as you see if. I will recommend adding them in hard to understand areas.

I haven't reviewed the core algorithms, and will do so in a subsequent review.

Also, this PR is in draft right now, could we elaborate on why it is so?

rahulramesha commented on app/client/src/pages/Editor/CanvasLayoutConversion/ConversionForm.tsx:
```diff
@@ -0,0 +1,177 @@
+  onCancel: () => void;
+};
+
+export type ConversionProps = {
```
Yes, Conversion form is a single component which renders different form states based on the properties supplied to it. So that a single component can render multiple different states based on the result of the hook, like useSnapshotFlow or useConversionForm etc.

rahulramesha commented on app/client/src/pages/Editor/CanvasLayoutConversion/InfoBlock.tsx:
```diff
@@ -0,0 +1,27 @@
+  info: string;
+};
+
+export const InfoBlock = (props: InfoBlockProps) => {
```
Noted

rahulramesha commented on app/client/src/pages/Editor/CanvasLayoutConversion/SnapShotButton.tsx:
```diff
@@ -0,0 +1,58 @@
+      isCloseButtonShown={false}
+      title={createMessage(USE_SNAPSHOT_HEADER)}
+      trigger={
+        <p className="font-semibold text-xs m-0 py-1 px-2 hover:bg-orange-200 cursor-pointer">
```
Changed this

rahulramesha commented:
@riodeuno, I have addressed most of the review comments.

I have currently kept it as draft because, this PR is using alpha version of design-system which is built on a open PR that i have on that repo. Once this passes QA, that PR will be merged to generate a stable version and update it here before merging to release.

rahulramesha commented:
/build-deploy-preview

rahulramesha commented:
/build-deploy-preview

riodeuno commented on app/client/src/pages/Editor/WidgetsEditor/index.tsx:
```diff
@@ -95,6 +111,22 @@ function WidgetsEditor() {
+      <div>
+        <Text
+          className="m-0 py-1 px-2 hover:bg-orange-200 cursor-pointer"
+          onClick={() => dispatch({ type: ReduxActionTypes.DELETE_SNAPSHOT })}
```
Let's use an action here instead of calling `{type: ReduxActionTypes.DELETE_SNAPSHOT}` individually.

riodeuno commented on app/client/src/pages/Editor/WidgetsEditor/index.tsx:
```diff
@@ -40,6 +52,10 @@ function WidgetsEditor() {
+  const isPreviewMode = useSelector(previewModeSelector);
+  const readableSnapShotDetails = useSelector(getReadableSnapShotDetails);
+
+  const shouldShowSnapShotBanner = !!readableSnapShotDetails && !isPreviewMode;
```
Are we using isPreviewMode anywhere else in this component? If not, can we have the `CanvasContainer` get it from the `useSelector`? The reason I'm saying is because, we will re-render every child of this component if the `isPreviewMode` value changes. Please ignore this if all the children anyway rely on `isPreviewMode` to change how they render, and in this case, we might want to create a task to pass the `isPreviewMode` from the parent to each of the children.

riodeuno commented on app/client/src/selectors/autoLayoutSelectors.tsx:
```diff
@@ -56,6 +71,39 @@ export const getParentOffsetTop = (widgetId: string) =>
+    const ExpirationInMilliseconds = 5 * 24 * millisecondsPerHour;
+    const timePassedSince = Date.now() - lastUpdatedDate.getTime();
+
+    const timeSince: string = dayjs
```
We need to be careful about how we're using dayjs. Ideally, I would want a utility library built on top of dayjs to provide helper functions which do all of these computations for us. The reason for this is that, we may find ourselves in a position to write the same computation in another feature, in which case, there might be duplication of code, and let's say, 2 years later, dayjs is no longer maintained, we won't be able to simply swap out the library.

riodeuno commented on app/client/src/utils/DSLConversions/autoToFixedLayout.ts:
```diff
@@ -0,0 +1,275 @@
+ */
+function getFixedCanvasWidget(
+  widgets: CanvasWidgetsReduxState,
+  canvasId: string,
```
`canvasWidgetId` would be more accurate. What do you think?

riodeuno commented on app/client/src/utils/DSLConversions/autoToFixedLayout.ts:
```diff
@@ -0,0 +1,275 @@
+  }
+
+  // Delete Canvas widget responsive properties
+  delete canvasWidget.flexLayers;
```
Are we not deleting the rest of the properties we've defined in `deletedResponsiveProperties`?

riodeuno commented on app/client/src/utils/DSLConversions/autoToFixedLayout.ts:
```diff
@@ -0,0 +1,275 @@
+} from "utils/autoLayout/flexWidgetUtils";
+import { DSLWidget } from "widgets/constants";
+
+const nonFlexLayerWidgets = ["MODAL_WIDGET"];
```
How do we define a `nonFlexLayerWidgets`? Is it like `detachFromLayout`?

riodeuno commented on app/client/src/utils/DSLConversions/autoToFixedLayout.ts:
```diff
@@ -0,0 +1,275 @@
+  for (const childId of canvasWidget.children || []) {
+    const currWidget = currWidgets[childId];
+
+    if (nonFlexLayerWidgets.includes(currWidget.type)) {
```
So, if we find a widget with `detachFromLayout` as a child of the Canvas Widget, I was wondering if we can take that as a way of defining this. If `detachFromLayout` doesn't work, then can we make this a property of the widget in the widget config? I'm only thinking out loud here, let me know.

riodeuno commented on app/client/src/utils/DSLConversions/autoToFixedLayout.ts:
```diff
@@ -0,0 +1,275 @@
+      leftColumn >
+      GridDefaults.DEFAULT_GRID_COLUMNS - HORIZONTAL_RESIZE_MIN_LIMIT
+    ) {
+      delete currWidgets[childId];
```
Are we deleting widgets from the DSL here? I am not sure I get what we're doing here.

riodeuno commented on app/client/src/utils/DSLConversions/fixedToAutoLayout.ts:
```diff
@@ -0,0 +1,729 @@
+  if (!dsl || !dsl.children) return dsl;
+
+  if (dsl.type === "CANVAS_WIDGET") {
+    return { ...getAutoCanvasWidget(dsl) };
```
We're spreading in the return of `getAutoCanvasWidget`, so I was wondering if we need to spread again?

riodeuno commented on app/client/src/utils/DSLConversions/fixedToAutoLayout.ts:
```diff
@@ -0,0 +1,729 @@
+import { DSLWidget } from "widgets/constants";
+
+const unHandledWidgets = ["LIST_WIDGET", "FORM_WIDGET", "MODAL_WIDGET"];
+const nonFlexLayerWidgets = ["MODAL_WIDGET"];
```
We're redefining this list in a few files. Could we do a few things here:
1. If we can, can we utilise an existing property which exists in the widget config?
2. Can we use one constant that we import for this?
3. How about adding this as a new configurations in widgets?

The rationale, is that when we add new widgets, we don't want multiple touch points

riodeuno reviewed (changes requested):
The logic behind the conversion, seems sound overall. There are quite a few hard to understand areas in the system, I'd like it if we can add comments explaining them.

rahulramesha commented:
/build-deploy-preview env=release

rahulramesha commented:
/build-deploy-preview env=release
